### ISSUE TITLE
erro

### ISSUE THREAD

**Issue by VascoLopo** [VascoLopo/ajuda#1]
import spacy
from sentence_transformers import SentenceTransformer
from collections import defaultdict
import json
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import itertools
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score
import numpy as np
import matplotlib.pyplot as plt
# Load the data
with open(r'C:\Users\Lopo\Downloads\project-10-at-2023-04-19-17-02-b9209ecb.json', encoding='utf-8') as f:
    data = json.load(f)

# Load the Portuguese model
nlp = spacy.load("pt_core_news_lg")
# Define the stopwords set
wordz = {"de", "a", "o", "que", "e", "do", "ao", "rt", "da", "em", "um", "para", "é", "com", "não", "uma",
             "os", "no", "se", "na", "por", "mais", "as", "dos", "como", "mas", "foi", "ao", "ele", "das",
             "tem", "à", "seu", "sua", "ou", "ser", "quando", "muito", "há", "nos", "já", "está", "eu",
             "também", "só", "pelo", "pela", "até", "isso", "ela", "entre", "era", "depois", "sem", "mesmo",
             "aos", "ter", "seus", "quem", "nas", "me", "esse", "eles", "estão", "você", "tinha", "foram",
             "essa", "num", "nem", "suas", "meu", "às", "minha", "têm", "numa", "pelos", "elas", "havia",
             "seja", "qual", "será", "nós", "tenho", "lhe", "deles", "essas", "esses", "pelas", "este", "fosse",
             "dele", "tu", "te", "vocês", "vos", "hei", "há", "havemos", "hão", "houve", "houvemos", "houveram",
             "houvera", "houvéramos", "haja", "hajamos", "hajam", "for", "is", "are", "this", "the", "and", "with",
             "to", "into", "on", "in", "of", "at", ".", ","}

# Define a function to preprocess the text
def preprocess_text(text):
    # Convert text to lowercase and remove words in WORDZ list
    words = text.lower().split()
    words = [word for word in words if word not in wordz]
    text = " ".join(words)

    # Tokenize the text
    doc = nlp(text)

    # Lemmatize and remove stopwords for each token
    lemmatized_tokens = [token.lemma_.lower() for token in doc if not token.is_stop]

    # Tag each token with its part of speech
    pos_tags = [(lemma_token, token.pos_) for token, lemma_token in zip(doc, lemmatized_tokens)]

    # Parse the sentence
    parse_tree = str(list(doc.sents)[0].root)

    return lemmatized_tokens, pos_tags, parse_tree

# Define a SentenceTransformer model
#model = SentenceTransformer('sentence-transformers/paraphrase-xlm-r-multilingual-v1')
model = SentenceTransformer('xlm-r-100langs-bert-base-nli-mean-tokens')
embeddings = [[] for _ in range(len(data))]
labels = [[] for _ in range(len(data))]
pos_tags = [[] for _ in range(len(data))]
parse_trees = [[] for _ in range(len(data))]

for i, item in enumerate(data):
    id = i
    annotations = item['annotations']
    lemmatized_tokens, pos_tags_list, parse_tree = preprocess_text(item['data']['text'])
    for annotation in annotations:
        result = annotation['result']
        if result:
            value = result[0]['value']
            choices = value.get('choices', [])
            if choices:
                choice = choices[0]
                sentence = item['data']['text']
                embedding = model.encode(sentence)
                pos_tag = pos_tags_list
                parse_trees[id].append(parse_tree)
                embeddings[id].append(embedding)
                labels[id].append(choice)
                pos_tags[id].append(pos_tag)
                parse_trees[id].append(parse_tree)

# Iterate over the embeddings, labels, POS tags, and parse trees
for i in range(len(embeddings)):
    print("ID:", i)
    print("Embeddings:", embeddings[i])
    print("Labels:", labels[i])
    print("POS tags:", pos_tags[i])
    print("Parse trees:", parse_trees[i])
    print("")

# Create an empty DataFrame with the desired columns
df = pd.DataFrame(columns=['ID', 'Embeddings', 'Labels', 'POS Tags', 'Parse Trees'])

# Iterate over the embeddings, labels, POS tags, and parse trees
for i in range(len(embeddings)):
    # Create a new row of data for each item
    row_data = {'ID': i, 'Embeddings': embeddings[i], 'Labels': labels[i], 'POS Tags': pos_tags[i], 'Parse Trees': parse_trees[i]}
    # Append the row to the DataFrame
    df = df.append(row_data, ignore_index=True)

# Set the display option to show all columns
pd.set_option('display.max_columns', None)

# Print the DataFrame
print(df.head(10))
sizes = set()
for i in range(len(embeddings)):
    for embedding in embeddings[i]:
        size = len(embedding)
        sizes.add(size)
print(sizes)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(embeddings, labels, test_size=0.25, random_state=42)

# Reshape the data to 2D arrays
X_train = [np.concatenate(x) for x in X_train]
X_test = [np.concatenate(x) for x in X_test]
y_train = [y[0] for y in y_train]
y_test = [y[0] for y in y_test]

# Fit a logistic regression model to the training data
clf = LogisticRegression(max_iter=1000)
clf.fit(X_train, y_train)

# Make predictions on the test data
y_pred = clf.predict(X_test)

# Compute the accuracy score, confusion matrix, precision, recall, and F1-score
accuracy = accuracy_score(y_test, y_pred)
cm = confusion_matrix(y_test, y_pred)
classification_report = classification_report(y_test, y_pred)

# Print the accuracy score, confusion matrix, precision, recall, and F1-score
print("Accuracy:", accuracy)
print("Confusion matrix:\n", cm)
print("Classification report:\n", classification_report)

# Plot the confusion matrix
plt.imshow(cm, interpolation='nearest', cmap=plt.cm.Blues)
plt.title("Confusion matrix")
plt.colorbar()
tick_marks = np.arange(len(set(y_test)))
plt.xticks(tick_marks, list(set(y_test)))
plt.yticks(tick_marks, list(set(y_test)))
thresh = cm.max() / 2.
for i, j in itertools.product(range(cm.shape[0]), range(cm.shape[1])):
    plt.text(j, i, format(cm[i, j], 'd'),
             horizontalalignment="center",
             color="white" if cm[i, j] > thresh else "black")
plt.tight_layout()
plt.ylabel('True label')
plt.xlabel('Predicted label')
plt.show()
# Create a meshgrid that covers the range of our training data
# convert X_train to a NumPy array
X_train = np.array(X_train)
xmin, xmax = X_train[:, 0].min() - 1, X_train[:, 0].max() + 1
ymin, ymax = X_train[:, 1].min() - 1, X_train[:, 1].max() + 1
xx, yy = np.meshgrid(np.arange(xmin, xmax, 0.1),
                     np.arange(ymin, ymax, 0.1))

# Create a dictionary that maps class labels to colors
color_dict = {"Populism": "red", "Non populism": "blue"}

# Map the class labels to colors using the dictionary
colors = [color_dict[label] for label in y_train]

# Create a scatter plot with the mapped colors
plt.scatter(X_train[:, 0], X_train[:, 1], c=colors)
plt.show()
